### ISSUE TITLE
Universal NSIS setup file for x64+arm64 (WoA) is broken

### ISSUE THREAD

**Issue by quanglam2807** [electron-userland/electron-builder#5461]
Universal NSIS setup file for x64+arm64 (Windows on ARM) is broken. Instead of installing the app, it wipes out the whole program directory, and copies only the `Uninstall ....exe` file (tested on x64 devices).

> Is this change ready? The PR that tried to revert this is closed: #4416 but when I tried to add arm64 support to my app, the universal-arch NSIS doesn't seem to work correctly on my x64 system. It only installs the `Uninstall {...}.exe` and ignores everything else. 

![Capture](https://user-images.githubusercontent.com/1548835/101584815-11a7d980-3a11-11eb-95ba-6b430f36f7c6.PNG)

_Originally posted by @quanglam2807 in https://github.com/electron-userland/electron-builder/issues/4228#issuecomment-741517647_

**Comment by gnattu** [comment]
Current `extractAppPackage.nsh ` is not correctly set, for example this part:

https://github.com/electron-userland/electron-builder/blob/aa3625d5f6b719b6a7634c4c7ba0aad1ffe7d3d6/packages/app-builder-lib/templates/nsis/include/extractAppPackage.nsh#L8-L13

Only  one of `ARM64` or `64` will be set, if arm64 target is defined, then `64` will not be set afterwards.

This also casues `arm64` targets not complie-able without setting `x64` as target, as `64` is undefined without `x64.`, then the script will only include `ia32` target in the target file.

Also this part:

https://github.com/electron-userland/electron-builder/blob/aa3625d5f6b719b6a7634c4c7ba0aad1ffe7d3d6/packages/app-builder-lib/templates/nsis/include/extractAppPackage.nsh#L49

`IsNativeARM64` will cause problems for devices not using arm CPU or running via emulation, in my test case, emulated x86 node on arm64 CPU, `IsNativeARM64` seems to return `false` in this script and asking me to provide an `x86_64` application.

**Comment by eengoron** [comment]
@mmaietta @lutzroeder @develar -- Are the any ideas here on how to fix this? Since we recently recently got universal builds for Apple Silicon working, this has become an issue on the Windows end. 

Thanks!

**Comment by lutzroeder** [comment]
@eengoron @quanglam2807 can you have a look at #4374 and #4228 which included changes to these files. The commit got reverted as it included an NSIS version change that regressed other scenarios. @dangeredwolf might  have some insights as well.

**Comment by quanglam2807** [comment]
@lutzroeder #4416 was not merged so #4228 changes remain intact and not reverted. 

I think we should try to fix `extractAppPackage.nsh` as @gnattu mentioned above. Unlike macOS, `electron-builder` opts to use universal binary on Windows so there's no way to build independent `arm64` and `x64` setup files in one build process. I think we should also make this consistent with macOS by adding `Arch.universal` support to Windows but it's not a simple task.

**Comment by lutzroeder** [comment]
@quanglam2807 makes sense and agree that fixing `extractAppPackage.nsh` seems more urgent to unblock the scenario. Not sure the universal approach will work. My understanding is Windows can have `x86`+`x64`+`arm64` as well? I'm not an expert reading NSIS scripts. Did you try the change @gnattu proposed?

**Comment by gnattu** [comment]
@lutzroeder What I did is a workaround for this situation so that we can have `arm64` and `x86+x64` nsis separated in two nsis files, but I did not figure out how to make a `x86+x64+arm64` universal package.

**Comment by lutzroeder** [comment]
@gnattu yes, this was more in reply to @quanglam2807 relating to Mac support. On Mac `universal` means `x64`+`arm64` and this might not make sense as a concept for Windows as there are multiple variants that could go into a single package.   Sounds like the right approach is to initially focus on a fix that allows merging all specified architectures into a single installer without breaking `x64` and have a separate complex pull request to figure out more detailed confirmation options (or sounds like you already have a way to do this).

**Comment by mmaietta** [comment]
@lutzroeder @gnattu @quanglam2807 Can you try using this script instead?
```
!macro extractEmbeddedAppPackage
  !ifdef COMPRESS
    SetCompress off
  !endif

  Var /GLOBAL packageArch
  
  !insertmacro identify_package
  !insertmacro compute_files_for_current_arch

  !ifdef COMPRESS
    SetCompress "${COMPRESS}"
  !endif

  !insertmacro decompress
  !insertmacro custom_files_post_decompression
!macroend

!macro identify_package 
  !ifdef APP_32
    StrCpy $packageArch "32"
  !endif
  !ifdef APP_64
    ${if} ${RunningX64}
    ${OrIf} ${IsNativeARM64}
      StrCpy $packageArch "64"
    ${endif}
  !endif
  !ifdef APP_ARM64
    ${if} ${IsNativeARM64}
      StrCpy $packageArch "ARM64"
    ${endif}
  !endif
!macroend

!macro compute_files_for_current_arch
  ${if} $packageArch == "ARM64"
    !ifdef APP_ARM64
      !insertmacro arm64_app_files
    !endif
  ${elseif} $packageArch == "64"
    !ifdef APP_64
      !insertmacro x64_app_files
    !endif
  ${else}
    !ifdef APP_32
      !insertmacro ia32_app_files
    !endif
  ${endIf}
!macroend

!macro custom_files_post_decompression
  ${if} $packageArch == "ARM64"
    !ifmacrodef customFiles_arm64
      !insertmacro customFiles_arm64
    !endif
  ${elseif} $packageArch == "64"
    !ifmacrodef customFiles_x64
      !insertmacro customFiles_x64
    !endif
  ${else}
    !ifmacrodef customFiles_ia32
      !insertmacro customFiles_ia32
    !endif
  ${endIf}
!macroend

!macro arm64_app_files
  File /oname=$PLUGINSDIR\app-arm64.${COMPRESSION_METHOD} "${APP_ARM64}"
!macroend

!macro x64_app_files
  File /oname=$PLUGINSDIR\app-64.${COMPRESSION_METHOD} "${APP_64}"
!macroend

!macro ia32_app_files
  File /oname=$PLUGINSDIR\app-32.${COMPRESSION_METHOD} "${APP_32}"
!macroend

!macro decompress
  !ifdef ZIP_COMPRESSION
    nsisunz::Unzip "$PLUGINSDIR\app-$packageArch.zip" "$INSTDIR"
  !else
    Nsis7z::Extract "$PLUGINSDIR\app-$packageArch.7z"
  !endif
!macroend

```

**Comment by lutzroeder** [comment]
@mmaietta can you Zip and share `extractAppPackage.nsh`? The patch did not apply.

**Comment by mmaietta** [comment]
@lutzroeder updated the comment to be the exact script instead of a patch.

**Comment by lutzroeder** [comment]
@mmaietta works for `x64`+`arm64` on both Windows x64 and Windows ARM. @gnattu did you try as well?

**Comment by matthiasg** [comment]
@mmaietta can make use of that script via some options (nsis/script or include) instead of patching electron-builder ?

**Comment by mmaietta** [comment]
@matthiasg you can include custom scripts via EB config:
```
    nsis: {
        include: "installer/win/nsis-installer.nsh"
    },
```

**Comment by matthiasg** [comment]
@mmaietta I tried that but it didn't work. It complained about duplicate macros. Also tried with script. Also tried to copy all existing scripts over and just overwriting etc.

**Comment by mmaietta** [comment]
@develar could you execute another pre-release cut to catch up with the latest changes on master please? 🙂

**Comment by mmaietta** [comment]
Merged in #5558 in v22.10.5 . Closing issue
